### PR TITLE
[feature](merge-cloud) Decouple rowset id generator and local rowsets gc implementation

### PR DESCRIPTION
**Pull request by platoneko**
## Proposed changes

Decouple rowset id generator and local rowsets garbage collection implementation. Instead, use pending rowset to implement local rowsets gc. Pending rowsets refer to those rowsets that are under construction, and have not been added to the tablet. BE storage engine will skip these pending rowsets during garbage collection.
Developers MUST add the rowset id to `PendingRowsetSet` before writing first row data to a local rowset, and hold returned `PendingRowsetGuard` during rowset construction before adding the rowset to tablet. Rowset id will be removed from `PendingRowsetSet` eventually when `PendingRowsetGuard` is destroyed.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by JackDrogon — CHANGES_REQUESTED** [review]

File: be/src/olap/data_dir.cpp [hunk 1]
```diff
@@ -627,74 +626,61 @@ Status DataDir::load() {
...
-            // could find the tablet, then skip check it
-            continue;
-        }
-        std::string data_dir_path =
-                io::Path(path).parent_path().parent_path().parent_path().parent_path();
-        DataDir* data_dir = StorageEngine::instance()->get_store(data_dir_path);
-        if (data_dir == nullptr) {
-            LOG(WARNING) << "could not find data dir for tablet path " << path;
+        auto tablet = _tablet_manager->get_tablet(tablet_id);
+        if (!tablet) {
```
[the remark this answers is not in the document]
Comment (platoneko) [reply be/src/olap/data_dir.cpp]: done

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

File: be/src/service/backend_service.cpp [hunk 2]
```diff
@@ -187,6 +188,8 @@ void _ingest_binlog(IngestBinlogArg* arg) {
         return;
     }
     RowsetId new_rowset_id = StorageEngine::instance()->next_rowset_id();
+    PendingRowsetGuard pending_rs_guard {new_rowset_id,
+                                         StorageEngine::instance()->pending_local_rowsets()};
```
Comment (gavinchou) [inline be/src/service/backend_service.cpp:192]: auto prg = StorageEngine::instance()->pending_local_rowsets()->add_pending_rowset(new_rs_id);

File: be/src/olap/rowset/pending_rowset_mixin.h [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+                        rowset_writer_context.fs->type() == io::FileSystemType::LOCAL)) {
+            // Only local rowset should register pending rowset here
+            _guard = PendingRowsetGuard {rowset_writer_context.rowset_id,
+                                         StorageEngine::instance()->pending_local_rowsets()};
+        }
+        return st;
+    }
+
+private:
+    PendingRowsetGuard _guard;
```
Comment (gavinchou) [inline be/src/olap/rowset/pending_rowset_mixin.h:51]: Add a comment for the guard.

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -1971,66 +1945,54 @@ void Tablet::reset_compaction(CompactionType compaction_type) {
...
 
-Status Tablet::create_rowset_writer(RowsetWriterContext& context,
-                                    std::unique_ptr<RowsetWriter>* rowset_writer) {
+Result<std::unique_ptr<RowsetWriter>> Tablet::create_pending_rowset_writer(
+        RowsetWriterContext& context, bool vertical) {
     context.rowset_id = StorageEngine::instance()->next_rowset_id();
     _init_context_common_fields(context);
-    return RowsetFactory::create_rowset_writer(context, false, rowset_writer);
+    std::unique_ptr<RowsetWriter> rowset_writer;
+    if (vertical) {
```
Comment (gavinchou) [inline be/src/olap/tablet.cpp:1986]: consider

```
struct RsMix : rowstwriter {
rs_guard; // expand rs_guard life span
};
```

**Comment by platoneko** [comment]
run buildall

**Comment by platoneko** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**